presto-main/src/main/java/com/facebook/presto/memory/RemoteNodeMemory.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import io.airlift.http.client.FullJsonResponseHandler.JsonResponse;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.HttpClient.HttpResponseFuture;
+import io.airlift.http.client.Request;
+import io.airlift.json.JsonCodec;
+import io.airlift.log.Logger;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.airlift.http.client.FullJsonResponseHandler.createFullJsonResponseHandler;
+import static io.airlift.http.client.Request.Builder.prepareGet;
+import static io.airlift.units.Duration.nanosSince;
+import static java.util.Objects.requireNonNull;
+import static java.util.Optional.empty;
+import static java.util.Optional.ofNullable;
+
+@ThreadSafe
+public class RemoteNodeMemory
+{
+    private static final Logger log = Logger.get(RemoteNodeMemory.class);
+
+    private final HttpClient httpClient;
+    private final URI memoryInfoUri;
+    private final JsonCodec<MemoryInfo> memoryInfoCodec;
+    private final AtomicReference<Optional<MemoryInfo>> memoryInfo = new AtomicReference<>(empty());
+    private final AtomicReference<Future<?>> future = new AtomicReference<>();
+    private final AtomicLong lastUpdateNanos = new AtomicLong();
+
+    public RemoteNodeMemory(HttpClient httpClient, JsonCodec<MemoryInfo> memoryInfoCodec, URI memoryInfoUri)
+    {
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+        this.memoryInfoUri = requireNonNull(memoryInfoUri, "memoryInfoUri is null");
+        this.memoryInfoCodec = requireNonNull(memoryInfoCodec, "memoryInfoCodec is null");
+    }
+
+    public Optional<MemoryInfo> getInfo()
+    {
+        return memoryInfo.get();
+    }
+
+    public void asyncRefresh()
+    {
+        if (nanosSince(lastUpdateNanos.get()).toMillis() > 1_000 && future.get() == null) {
```

Comment:
They do, but there are always bugs and things like the server "slow rolling" the client